### PR TITLE
Add DIVISIBLE_BY operator to NumberRule

### PR DESCRIPTION
**Pull request by jetfreakk**
## Description

This change add DIVISIBLE_BY operator  to the NumberRule.

## Related Issue

None

## Motivation and Context

Provides more Operators to regola rules

## How Has This Been Tested?

Unit Tests and Custom Implementation for use case

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/java/com/adobe/abp/regola/rules/NumberRule.java [hunk 1]
```diff
@@ -47,6 +47,10 @@ public NumberRule(Executor executor, String key, Operator operator, V value) {
 
     @Override
     Result check(V fact) {
+        if (fact == null || getValue() == null) {
+            return Result.INVALID;
```
Comment (jetfreakk) [inline src/main/java/com/adobe/abp/regola/rules/NumberRule.java:51]: Return early if fact or value is missing

Comment (sic2) [reply src/main/java/com/adobe/abp/regola/rules/NumberRule.java:51]: This is already covered by:

```java
<F> Result checkFact(F fact, BiPredicate<F, T> predicate) {
        return Optional.ofNullable(fact)
                .filter(f -> Objects.nonNull(getValue()))
```

File: src/main/java/com/adobe/abp/regola/rules/NumberRule.java [hunk 2]
```diff
@@ -58,11 +62,25 @@ Result check(V fact) {
                 return checkFact(fact, (f, value) -> comparator.compare(f, getValue()) < 0);
             case LESS_THAN_EQUAL:
                 return checkFact(fact, (f, value) -> comparator.compare(f, getValue()) <= 0);
+            case MODULO:
+                return checkFact(fact, (f, value) -> f.doubleValue() % getValue().doubleValue() == 0);
```
Comment (sic2) [inline src/main/java/com/adobe/abp/regola/rules/NumberRule.java:66]: Modulo arithmetic and divisibility is primarily allowed for integers only (there are ways to perform mod on reals, but it is a bit niche as far as I am aware). 

So while `%` will work on reals (double, float in Java), it is not as conventional in mathematics and a bit surprising.

File: src/test/java/com/adobe/abp/regola/rules/NumberRuleDoubleTest.java [hunk 3]
```diff
@@ -471,6 +471,97 @@ void integerFactIsNotValid() {
...
+        }
+
+        @Test
+        @DisplayName("evaluate as invalid if (integer) fact does not match")
+        void integerFactIsNotValid() {
+            when(resolver.resolveFact(RULE_KEY))
+                    .thenReturn(CompletableFuture.supplyAsync(() -> 7));
+
+            RuleTestUtils.evaluateAndTestWithAnyTypeValue(rule, resolver, ruleResultBuilderMixedTypes, Result.INVALID, 7, 3.0);
+        }
```
Comment (sic2) [inline src/test/java/com/adobe/abp/regola/rules/NumberRuleDoubleTest.java:562]: We will need to test negative numbers and the zero too.

**Review by sic2 — COMMENTED** [review]
Good contribution.

As discussed offline, we will be moving from `MODULO` to `DIVISIBLE_BY`, since `MODULO` can be thought of as a function of `(number, number) --> number`, while we need `(number, number) --> bool`.

The main comment on this PR is that we should restrict ourselves to integers and either not support doubles or say that we do not guarantee behaviour.

File: src/main/java/com/adobe/abp/regola/rules/NumberRule.java (hunk 2 above)
Comment (jetfreakk) [reply src/main/java/com/adobe/abp/regola/rules/NumberRule.java:66]: So should we explicitly restrict the DIVISIBLE_BY operation to integer types?

Comment (sic2) [reply src/main/java/com/adobe/abp/regola/rules/NumberRule.java:66]: I think so.

File: src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java [hunk 4]
```diff
@@ -469,6 +469,107 @@ void doubleFactIsNotValid() {
         }
     }
 
+    @Nested
+    @DisplayName("divisible-by should")
+    class DivisibleBy {
+
+        private ValuesRuleResult.RuleResultBuilder<Integer> ruleResultBuilder;
+        private ValuesRuleResult.RuleResultBuilder<Object> ruleResultBuilderMixedTypes;
```
Comment (sic2) [inline src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java:477]: This is never used. Delete.

File: src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java [hunk 5]
```diff
@@ -469,6 +469,107 @@ void doubleFactIsNotValid() {
...
+                r.expectedValue = 3;
+            });
+            ruleResultBuilderMixedTypes = ValuesRuleResult.builder().with(r -> {
+                r.type = RuleType.NUMBER.getName();
+                r.operator = Operator.DIVISIBLE_BY;
+                r.key = RULE_KEY;
+                r.expectedValue = 3;
+            });
+        }
+
```
Comment (sic2) [inline src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java:497]: Add test to make sure we cannot divide by zero

File: README.md [hunk 6]
```diff
@@ -233,6 +233,8 @@ Integer-Double comparisons between the rule's value and the data provided by the
 | 7          | CONTAINS           | [ 6, 8]      | INVALID |
 | 7          | CONTAINS           | [ 6, 7.0, 8] | INVALID |
 | 7.0        | CONTAINS           | [ 6, 7.0, 8] | VALID   |
+| 7.0        | DIVISIBLE_BY       | 7.0          | VALID   |
+| 7.0        | DIVISIBLE_BY       | 8.0          | INVALID |
```
Comment (sic2) [inline README.md:237]: Update to use integers

File: src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java [hunk 7]
```diff
@@ -469,6 +469,107 @@ void doubleFactIsNotValid() {
...
+        @Test
+        @DisplayName("evaluate as valid if (negative) fact is perfectly divisible by positive divisor")
+        void negativeFactPositiveDivisorIsDivisible() {
+            when(resolver.resolveFact(RULE_KEY))
+                    .thenReturn(CompletableFuture.supplyAsync(() -> -9));
+
+            RuleTestUtils.evaluateAndTest(rule, resolver, ruleResultBuilder, Result.VALID, -9, 3);
+        }
+    }
```
Comment (sic2) [inline src/test/java/com/adobe/abp/regola/rules/NumberRuleIntegerTest.java:571]: Tests on longs are missing

File: src/main/java/com/adobe/abp/regola/rules/NumberRule.java [hunk 8]
```diff
@@ -58,11 +58,30 @@ Result check(V fact) {
                 return checkFact(fact, (f, value) -> comparator.compare(f, getValue()) < 0);
             case LESS_THAN_EQUAL:
                 return checkFact(fact, (f, value) -> comparator.compare(f, getValue()) <= 0);
+            case DIVISIBLE_BY:
+                if (!(fact instanceof Integer || fact instanceof Long) ||
```
Comment (sic2) [inline src/main/java/com/adobe/abp/regola/rules/NumberRule.java:62]: Not all branches are tested.

Screenshot 2024-11-25 at 17 15 20

**Review by sic2 — COMMENTED** [review]
Changes look good. We need more tests however.

**Review by sic2 — APPROVED** [review]
@jetfreakk thank you for this contributon. I have reviewed these changes and discussed them with you offline. They look good. I have pushed some minor adjustments on top of yours, but all looks fine to me. Approving.
